feat: Support adding force directly to simulation.

b-butler opened the pull request:
## Description
This makes the force a Compute on the Python side so it can be querried for force and energy even when not applied to the system.
<!-- Describe your changes in detail. -->

## Motivation and context
This was motivated by a use case for querying a non-existent force by @Charlottez112.

## How has this been tested?
~Has not been tested currently, though I plan to add tests soon.~ Tests added.

## Change log

<!-- Propose a change log entry. -->
```
Added
- MD forces can now be added to the Operations list allowing for force, energy, virial querying without the forces effecting the system.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

b-butler committed:
feat: Support adding force directly to simulation.

This makes the force a Compute on the Python side so it can be querried
for force and energy even when not applied to the system.

b-butler committed:
test: Add a test for Force objects as computes

b-butler committed:
feat: Allow for intra operation list dependencies.

We will no longer call `~._attach` twice in this case. This introduces
a new internal API `~._attach_hook` and `~._detach_hook` which provides
customization upon attaching and detaching while the calling methods
keep track of the number of times they have been called as a way to know
the number of consumers that an object has.

joaander commented on hoomd/data/syncedlist.py:
```diff
@@ -199,7 +199,12 @@ def _attach_value(self, value, raise_if_added=True):
+            # attached first. Assuming the code for adding a simulation to an
+            # operation is correct, this is safe.
+            if not value._attached:
+                value._attach()
```
Could we test this in a unit test that adds lj to both the `integrator.forces` and `operations.computes`? I suppose the test could use internal members to ensure that the `_cpp_obj` is the same for both and/or that it doesn't change when adding it the 2nd time.

Should we also test unsetting the integrator? Need to make sure this doesn't forcibly detach the force that is present in `computes`.

joaander reviewed (commented):
Looks good! I have some questions on testing the dependent add fix.

b-butler committed:
refactor: Improve error handling of attaching/detaching

We now ensure we clean up after ourselves if an error is raised when
attaching/detaching or removing.

b-butler committed:
refactor: Switch neighborlist exclusions C++ getter to list

This is necessary to prevent errors in the collections logic as we
expect a tuple to be always of the same size, and to have a specific
style of type validator.

b-butler commented:
Wait on reviewing this. I still have to fix one problem in adding an object to two `SyncedList` objects.

b-butler committed:
refactor: SyncedList method names to avoid confusion

Remove attach and detach from the names as these method may or may not
attach/detach and may or may not add/remove.

b-butler committed:
refactor: Allow for an object to be in two SyncedLists

This removes the pre-sync checks for inclusion in two lists and errors
on syncing if any item is already added to another simulation.

b-butler committed:
test: Test adding a force to computes and integrator

Also updates existing tests to pass now than an item can be in more than
one synced list at a time (this defers some error messages which we test
for).

b-butler committed:
test: Update SyncedList tests.

b-butler commented:
Good to review again.

joaander reviewed (approved):
Thanks! This looks great!
